Extend binlogapi

andreas-bok-sociomantic opened the pull request:
Adds two utility function to the binlog api

`drizzle_binlog_event_type_str`
Returns the enum name of a binlog event type as const char*

`drizzle_binlog_get_filename`
The added function allows the client to retrieve the filename of a binlog specified by index.
If the index is -1, the filename of the latest binlog is returned

andreas-bok-sociomantic committed:
Adds two function to the binlog api

`drizzle_binlog_event_type_str`
Returns the enum name of a binlog event type
as const char*

`drizzle_binlog_get_filename`
The added function allows the client to retrieve
the filename of a binlog specified by index.
If the index is -1, the filename of the latest
binlog is returned

ben-palmer-sociomantic commented on libdrizzle-5.1/binlog.h:
```diff
@@ -183,6 +183,32 @@ const unsigned char *drizzle_binlog_event_raw_data(drizzle_binlog_event_st *even
+ * @param[in] con Drizzle structure previously initialized with
+ *  drizzle_create() or drizzle_clone().
+ * @param[in] file_index index of the binlog to retrieve.
+ * @return A binlog filename or an empty string
```
Is this actually a return value?

andreas-bok-sociomantic replied:
Ups, nope. didn't update the doc after modifying the function

ben-palmer-sociomantic commented on libdrizzle/binlog.cc:
```diff
@@ -398,3 +398,79 @@ drizzle_return_t drizzle_state_binlog_read(drizzle_st *con)
+  drizzle_result_buffer(result);
+  int64_t row_count = drizzle_result_row_count(result);
+
+  if (file_index < -1 || file_index >= row_count )
```
Should this check be done at the start of the method?

andreas-bok-sociomantic replied:
No. It depends on before buffering the whole result in the previous line, which again depends on a successful query.

ben-palmer-sociomantic replied:
Right, makes sense.

ben-palmer-sociomantic commented:
Finished my first review.

andreas-bok-sociomantic committed:
Update rst documentation for binlogapi

Adds documentation for the functions:
 - drizzle_binlog_event_type_str
 - drizzle_binlog_get_filename

andreas-bok-sociomantic commented:
Updated based on comments from @ben-palmer-sociomantic 
Also updated rst documentation for the new functions 

```diff
diff --git a/docs/api/binlog.rst b/docs/api/binlog.rst
index ef014ae..b16ceb5 100644
--- a/docs/api/binlog.rst
+++ b/docs/api/binlog.rst
@@ -153,3 +153,25 @@ Functions
    :param event: The event from the binlog stream
    :returns: The length of the raw event data
 
+.. c:function:: const char *drizzle_binlog_event_type_str(drizzle_binlog_event_types_t event_type)
+
+   Get the event type for the binlog event as string
+
+   :param event_type: A binlog event type
+   :returns: The event type of the binlog event as string
+
+.. c:function:: void drizzle_binlog_get_filename(drizzle_st *con, char *filename, int file_index)
+
+   Get the name of a binlog-file
+
+   Queries the database for a list of binlog files and copies the filename to
+   the passed buffer
+   If the file_index is invalid or no binlog files exist filename will contain an
+   empty string
+   A valid file_index is in the range [-1 to (number of binlog files -1)]
+
+   The memory (de)allocation of the filename buffer must be done by the client
+
+   :param con: Drizzle structure previously initialized with drizzle_create() or drizzle_clone()
+   :param filename: Buffer to copy filename to
+   :param file_index: Index of the binlog to retrieve
diff --git a/libdrizzle-5.1/binlog.h b/libdrizzle-5.1/binlog.h
index 9905489..a5f3a58 100644
--- a/libdrizzle-5.1/binlog.h
+++ b/libdrizzle-5.1/binlog.h
@@ -193,18 +193,20 @@ DRIZZLE_API
 const char *drizzle_binlog_event_type_str(drizzle_binlog_event_types_t event_type);
 
 /**
- * Get a binlog filename
+ * Get the name of a binlog-file
  *
- * If file_index is -1 the filename of binlog currently in use is returned.
- * If no binlog files exist an empty string is returned.
- * Note that file_index must be in the range [-1 to (number of binlog files -1)]
+ * Queries the database for a list of binlog files and copies the filename to
+ * the passed buffer
+ * If the file_index is invalid or no binlog files exist filename will contain an
+ * empty string.
+ * A valid file_index is in the range [-1 to (number of binlog files -1)]
  *
- * The memory handling of the filename buffer must be handled by the client
+ * The memory (de)allocation of the filename buffer must be done by the client
  *
  * @param[in] con Drizzle structure previously initialized with
  *  drizzle_create() or drizzle_clone().
+ * @param[in,out] filename buffer to copy filename to
  * @param[in] file_index index of the binlog to retrieve.
- * @return A binlog filename or an empty string
  */
 DRIZZLE_API
 void drizzle_binlog_get_filename(drizzle_st *con, char *filename, int file_index);
```

andreas-bok-sociomantic commented:
rebased on master

ben-palmer-sociomantic commented on libdrizzle/binlog.cc:
```diff
@@ -398,3 +398,79 @@ drizzle_return_t drizzle_state_binlog_read(drizzle_st *con)
+    }
+}
+
+void drizzle_binlog_get_filename(drizzle_st *con, char *filename, int file_index)
```
I might be missing something here but how does the caller know how big `filename` has to prevent the `memcpy` overwriting other memory?

andreas-bok-sociomantic replied:
Right, I will move allocation inside the function and change the comment to be that the caller is responsible for deallocating the buffer

ben-palmer-sociomantic commented:
The first line of the commit message for the commit `Extend binlogapi` should be `Adds two functions to the binlog api:`.

andreas-bok-sociomantic commented:
- update based on comments from @ben-palmer-sociomantic 
- rebased on master

It should go into v5.3 since it extends the api. How should I go about making the new upstream v5.3.x branch?

```diff
diff --git a/libdrizzle-5.1/binlog.h b/libdrizzle-5.1/binlog.h
index a5f3a58..0d4582a 100644
--- a/libdrizzle-5.1/binlog.h
+++ b/libdrizzle-5.1/binlog.h
@@ -201,7 +201,8 @@ const char *drizzle_binlog_event_type_str(drizzle_binlog_event_types_t event_typ
  * empty string.
  * A valid file_index is in the range [-1 to (number of binlog files -1)]
  *
- * The memory (de)allocation of the filename buffer must be done by the client
+ * The filename parameter is allocated by the function and needs to be
+ * freed by the application when finished with.
  *
  * @param[in] con Drizzle structure previously initialized with
  *  drizzle_create() or drizzle_clone().
@@ -209,7 +210,7 @@ const char *drizzle_binlog_event_type_str(drizzle_binlog_event_types_t event_typ
  * @param[in] file_index index of the binlog to retrieve.
  */
 DRIZZLE_API
-void drizzle_binlog_get_filename(drizzle_st *con, char *filename, int file_index);
+void drizzle_binlog_get_filename(drizzle_st *con, char **filename, int file_index);
 
 #ifdef __cplusplus
 }
diff --git a/libdrizzle/binlog.cc b/libdrizzle/binlog.cc
index f0edc19..b082dfe 100644
--- a/libdrizzle/binlog.cc
+++ b/libdrizzle/binlog.cc
@@ -444,7 +444,7 @@ const char *drizzle_binlog_event_type_str(drizzle_binlog_event_types_t event_typ
     }
 }
 
-void drizzle_binlog_get_filename(drizzle_st *con, char *filename, int file_index)
+void drizzle_binlog_get_filename(drizzle_st *con, char **filename, int file_index)
 {
   drizzle_return_t driz_ret;
   drizzle_result_st *result = drizzle_query(con, "SHOW BINARY LOGS", 0, &driz_ret);
@@ -462,14 +462,18 @@ void drizzle_binlog_get_filename(drizzle_st *con, char *filename, int file_index
     drizzle_log_error(con, "%s Invalid binlog file index %d", __func__, file_index );
   }
 
+  *filename = (char*) malloc(1);
+  *filename[0] = '\0';
+
   if (row_count > 0)
   {
     int row_index = file_index == -1 || file_index >= row_count ?
                             row_count - 1 : file_index;
 
     drizzle_row_t row = drizzle_row_index(result, row_index);
-    memcpy(filename, row[0], strlen(row[0]));
-    filename[strlen(row[0])] = '\0';
+    *filename =( char*)realloc(*filename, strlen(row[0]));
+    memcpy(*filename, row[0], strlen(row[0]));
+    *filename[strlen(row[0])] = '\0';
   }
 
   drizzle_result_free(result);
```

andreas-bok-sociomantic commented:
blocked until upstream v5.3.x branch has been created and set as base for the PR

ben-palmer-sociomantic commented:
> blocked until upstream v5.3.x branch has been created and set as base for the PR

Created upstream v5.3.x branch.

andreas-bok-sociomantic commented:
Unblocked. Changed base branch to v5.3.x

andreas-bok-sociomantic commented:
Changed base branch to v5.x.x

ben-palmer-sociomantic commented:
LGTM

ben-palmer-sociomantic commented:
This pull request has been rebased via `git hub pull rebase`. Original pull request HEAD was f29673be81f91ef37f15c64c95ba48badf04bc36, new (rebased) HEAD is da433d00ed13822b9304851d4fd3fdcee921f60d
